libyul/optimiser/StackToMemoryMover.cpp

```diff
@@ -98,68 +117,160 @@ m_nameDispenser(_context.dispenser)
+		});
+		// Replace original function by a call to the new function and an assignment to the return variable from memory.
+		_functionDefinition.body = Block{_functionDefinition.location, move(memoryVariableInits)};
+		_functionDefinition.body.statements.emplace_back(ExpressionStatement{
+			_functionDefinition.location,
+			FunctionCall{
+				_functionDefinition.location,
+				Identifier{_functionDefinition.location, newFunctionName},
+				stackArguments | ranges::views::transform([&](TypedName const& _arg) {
+					return Expression{Identifier{
+						_functionDefinition.location,
+						_arg.name
+					}};
+				}) | ranges::to<vector<Expression>>
+			}
+		});
+		_functionDefinition.body.statements.emplace_back(Assignment{
+			_functionDefinition.location,
+			{Identifier{_functionDefinition.location, _functionDefinition.returnVariables.front().name}},
+			make_unique<Expression>(generateMemoryLoad(
+				m_context.dialect,
+				_functionDefinition.location,
+				*m_memoryOffsetTracker(_functionDefinition.returnVariables.front().name)
+			))
+		});
+		return;
+	}
+
+	if (!memoryVariableInits.empty())
+	{
+		memoryVariableInits += move(_functionDefinition.body.statements);
+		_functionDefinition.body.statements = move(memoryVariableInits);
+	}
+
+	_functionDefinition.returnVariables = _functionDefinition.returnVariables | ranges::views::filter([&](TypedName const& _name){
+		return !m_memoryOffsetTracker(_name.name);
+	}) | ranges::to<TypedNameList>;
 }
 
 void StackToMemoryMover::operator()(Block& _block)
 {
 	using OptionalStatements = std::optional<vector<Statement>>;
-	auto rewriteAssignmentOrVariableDeclaration = [&](
+
+	auto rewriteAssignmentOrVariableDeclarationLeftHandSide = [this](
 		auto& _stmt,
-		auto const& _variables
+		auto& _lhsVars
 	) -> OptionalStatements {
 		using StatementType = decay_t<decltype(_stmt)>;
-		if (_stmt.value)
-			visit(*_stmt.value);
-		bool leftHandSideNeedsMoving = util::contains_if(_variables, [&](auto const& var) {
-			return m_memoryOffsetTracker(var.name);
-		});
-		if (!leftHandSideNeedsMoving)
-			return {};
-
-		langutil::SourceLocation loc = _stmt.location;
 
-		if (_variables.size() == 1)
+		langutil::SourceLocation  loc = _stmt.location;
+		if (_lhsVars.size() == 1)
 		{
-			optional<YulString> offset = m_memoryOffsetTracker(_variables.front().name);
-			yulAssert(offset, "");
-			return generateMemoryStore(
-				m_context.dialect,
-				loc,
-				*offset,
-				_stmt.value ? *std::move(_stmt.value) : Literal{loc, LiteralKind::Number, "0"_yulstring, {}}
-			);
+			optional<YulString> offset = m_memoryOffsetTracker(_lhsVars.front().name);
+			if (offset)
+				return generateMemoryStore(
+					m_context.dialect,
+					loc,
+					*offset,
+					_stmt.value ? *std::move(_stmt.value) : Literal{loc, LiteralKind::Number, "0"_yulstring, {}}
+				);
+			else
+				return {};
 		}
+		FunctionCall const* functionCall = get_if<FunctionCall>(_stmt.value.get());
+		yulAssert(functionCall, "");
```

Comment:
Why is this true?

Reply:
Because there are multi-assignments or variable declarations in yul unless the right hand side is a function call.

Reply:
(I.e. the case of single-assignments is entirely covered in the ``_lhsVars.size() == 1`` case. ``_lhsVars.empty()`` cannot happen either, so here we are in the case ``_lhsVars.size() > 1``, which implies the right-hand-side being a function call)

Reply:
Actually, @bshastry just came up with an example in which this is actually false - that is ``let x, y, z`` without right-hand-side :-). I think we will actually split those during other optimizations, but I'll still change this, to handle those gracefully as well.

Reply:
Updated.